### PR TITLE
Remove Makefile and migrate to NPM scripts

### LINKED ISSUE #1121: remove Makefile
Many of the existing targets have already been migrated to package.json.  The release.sh script will probably need a little tweaking.

### PR DESCRIPTION
**Pull request by ylemkimon**
Fixes #1121.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -39,6 +39,7 @@
     "jspngopt": "^0.2.0",
     "less": "~2.7.1",
     "less-loader": "^4.0.5",
+    "mkdirp": "^0.5.1",
```
Comment (ylemkimon) [inline package.json]: For cross-platform `mkdir`

Comment (ylemkimon) [inline package.json]: For cross-platform `mkdir`

File: package.json [hunk 2]
```diff
@@ -62,13 +63,18 @@
...
     "clean": "rm -rf build/* node_modules/",
     "clean-install": "npm run clean && npm i",
-    "test": "check-dependencies && npm run lint && npm run flow && npm run jest",
-    "verify-screenshots": "check-dependencies && dockers/Screenshotter/screenshotter.sh --verify",
-    "start": "check-dependencies && webpack-dev-server --hot --config webpack.dev.js",
-    "build": "check-dependencies && rimraf build/* && webpack",
+    "test": "npm run prestart && npm run lint && npm run flow && npm run jest",
+    "screenshots": "npm run prestart && dockers/Screenshotter/screenshotter.sh",
+    "verify-screenshots": "npm run screenshots -- --verify",
+    "prestart": "check-dependencies && cd src && node unicodeMake.js >unicodeSymbols.js",
```
Comment (ylemkimon) [inline package.json]: As generating `unicodeSymbols.js` is cheap, it does before every tests and builds.

File: release.sh [hunk 3]
```diff
@@ -104,7 +104,7 @@ git checkout --detach
 
 # Build generated files and add them to the repository (for bower)
 git clean -fdx build dist
-make setup dist
+npm run dist
```
Comment (ylemkimon) [inline release.sh]: Should this command run when `--dry-run`?

File: Makefile [hunk 4]
```diff
@@ -1,95 +0,0 @@
...
-
-zip: build/katex.tar.gz build/katex.zip
-
-compress: build/katex.min.js build/katex.min.css
-	@JSSIZE=`gzip -c build/katex.min.js | wc -c`; \
-	CSSSIZE=`gzip -c build/katex.min.css | wc -c`; \
-	TOTAL=`echo $${JSSIZE}+$${CSSSIZE} | bc`; \
-	printf "Minified, gzipped js:  %6d\n" "$${JSSIZE}"; \
-	printf "Minified, gzipped css: %6d\n" "$${CSSSIZE}"; \
-	printf "Total:                 %6d\n" "$${TOTAL}"
```
Comment (ylemkimon) [inline Makefile]: I'm not sure how to do this easily in the NPM scripts.

File: Makefile [hunk 5]
```diff
@@ -1,95 +0,0 @@
...
-	$(NPM) test
-
-coverage: $(NIS)
-	$(NPM) run coverage
-
-PERL=perl
-PYTHON=$(shell python2 --version >/dev/null 2>&1 && echo python2 || echo python)
-
-metrics:
-	cd metrics && $(PERL) ./mapping.pl | $(PYTHON) ./extract_tfms.py | $(PYTHON) ./extract_ttfs.py | $(PYTHON) ./format_json.py --width > ../src/fontMetricsData.js
```
Comment (ylemkimon) [inline Makefile]: This will be moved to the fonts repo by #1134.

File: package.json [hunk 6]
```diff
@@ -62,13 +63,18 @@
...
     "clean": "rm -rf build/* node_modules/",
     "clean-install": "npm run clean && npm i",
-    "test": "check-dependencies && npm run lint && npm run flow && npm run jest",
-    "verify-screenshots": "check-dependencies && dockers/Screenshotter/screenshotter.sh --verify",
-    "start": "check-dependencies && webpack-dev-server --hot --config webpack.dev.js",
-    "build": "check-dependencies && rimraf build/* && webpack",
+    "test": "npm run prestart && npm run lint && npm run flow && npm run jest",
+    "screenshots": "npm run prestart && dockers/Screenshotter/screenshotter.sh",
+    "verify-screenshots": "npm run screenshots -- --verify",
+    "prestart": "node check-node-version.js && check-dependencies && cd src && node unicodeMake.js >unicodeSymbols.js",
```
Comment (ylemkimon) [inline package.json]: As generating `unicodeSymbols.js` is cheap, it does before every tests and builds.

File: Makefile (hunk 5 above)
Comment (kevinbarabash) [reply Makefile]: I'm going to hold of on merging this until #1134 has been merged.

File: package.json (hunk 6 above)
Comment (kevinbarabash) [reply package.json]: One less thing to remember to do. :)

File: .arclint [hunk 7]
```diff
@@ -2,7 +2,7 @@
   "linters": {
     "katex-linter": {
       "type": "script-and-regex",
-      "script-and-regex.script": "make lint || true",
+      "script-and-regex.script": "npm run lint || true",
```
Comment (kevinbarabash) [inline .arclint]: Thanks for updating this file.  We use phabricator/arcanist internally, but I can't remember the last time I saw a phabricator review for KaTeX.

File: release.sh (hunk 3 above)
Comment (kevinbarabash) [reply release.sh]: I think so.  It won't do anything other than echo the command to the shell which is what we want in the dry run.

File: package.json [hunk 8]
```diff
@@ -62,13 +63,18 @@
...
+    "start": "webpack-dev-server --hot --config webpack.dev.js",
+    "build": "npm run prestart && rimraf build/* && webpack",
+    "dist": "npm test && npm run build && npm run dist:copy && npm run dist:zip && npm run dist:dist",
+    "dist:copy": "cd build && mkdirp katex && cp -r katex.js katex.min.js katex.css katex.min.css contrib fonts ../README.md katex",
+    "dist:zip": "cd build && tar czf katex.tar.gz katex/ && zip -rq katex.zip katex/",
+    "dist:dist": "rimraf dist/ && cp -r build/katex/ dist/",
```
Comment (kevinbarabash) [inline package.json]: I like this organization.  Could you do something similar for `test`?

File: release.sh (hunk 3 above)
Comment (ylemkimon) [reply release.sh]: @kevinbarabash Do you mean it should actually build the KaTeX or not?

Comment (kevinbarabash) [reply release.sh]: Sorry, the real command should not run, it should jus echo what the command to the console.

**Pull request merged**